### PR TITLE
execution details persistence

### PR DESCRIPTION
**Pull request by JoooJooo**
1. Before initiating Execution DAG, an entry is made in MongoDB with executionID as Key and intial-payload in the the documentDB.
2. Created saveExecutionDetails and fetchExecutionDetails in the shpeherd client, so that Customer serivices can fetch and change the document on given exeutionId.

### CODE REVIEW

**Comment by JoooJooo** [comment]
image

**Comment by JoooJooo** [comment]
1st document is the when DAG is about to start to be executed.
2nd Document is after one of the node made some update on it .

File: shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/PayLoad.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
+package com.devsda.platform.shepherdclient.impl;
+
+import java.util.Date;
+
+public class PayLoad {
```
Comment (devvsda) [inline shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/PayLoad.java]: This makes document  rigid. Payload should be Map<String, Object>. Wanted to talk more on this. Lets discuss offline.

File: shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/ShepherdClientTest.java [hunk 2]
```diff
@@ -94,4 +95,22 @@ public void executeEndpointTest() throws Exception {
         ObjectMapper objectMapper = new ObjectMapper();
         System.out.println(objectMapper.writeValueAsString(executeEndpointResponse));
     }
+
+    @Test public void saveDocumentTest() {
```
Comment (devvsda) [inline shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/ShepherdClientTest.java]: +1 to see UTs for Document sharing usecase.

File: shepherd-model/src/main/java/com/devsda/platform/shepherd/model/DataSourceDetails.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    private ArrayList<String> clusters;
+
+    private String path;
+    private String params;
+
+    public String getDb_name() {
+        return db_name;
+    }
+
+    public void setDb_name(String db_name) {
```
Comment (devvsda) [inline shepherd-model/src/main/java/com/devsda/platform/shepherd/model/DataSourceDetails.java]: Don't use Snake case convention. Let it be there in Python.:-D. Don't ruin the beauty of Java. :-)

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java [hunk 4]
```diff
@@ -31,12 +35,11 @@
     @Inject
     public ResourceHelper resourceHelper;
 
-
     @POST
     @Path(ShephardConstants.Resources.ENDPOINT)
     @Produces(MediaType.APPLICATION_JSON)
     @Consumes(MediaType.APPLICATION_JSON)
-    public Response executeWorkflow(@NotNull ExecuteWorkflowRequest executeWorkflowRequest) {
+    public Response executeWorkflow(ExecuteWorkflowRequest executeWorkflowRequest) {
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java]: Not intentionally removed.

[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java]: corrected

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/HealthCheckReources.java [hunk 5]
```diff
@@ -17,6 +18,7 @@
 @Produces(MediaType.APPLICATION_JSON)
 @Consumes(MediaType.APPLICATION_JSON)
 public class HealthCheckReources {
+    private static final Logger log = LoggerFactory.getLogger(HealthCheckReources.class);
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/HealthCheckReources.java]: Sure

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ResourceHelper.java [hunk 6]
```diff
@@ -24,7 +24,7 @@ public ShepherdResponse createShepherdResponse(ResourceName resourceName, Map<St
         shepherdResponse.setResponseData(resourceResponse);
 
         try {
-            shepherdResponse.setServerName(InetAddress.getLocalHost().getHostName());
+            shepherdResponse.setServerName(InetAddress.getLocalHost().getHostName()+"ASHUTOSH ");
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ResourceHelper.java]: During testing, removed now.

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/service/ExecuteWorkflowService.java [hunk 7]
```diff
@@ -32,6 +44,8 @@
     @Inject
     private WorkflowOperationDao workflowOperationDao;
 
+    @Inject
+    private MongoDBService mongoDBService;
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/service/ExecuteWorkflowService.java]: changed

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/service/MongoDBService.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    public boolean insertExecutionDetails(Integer executionId, Map<String, Object> initialPayload) {
+        this.mongoClient = getMongoClient();
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            mapper.configure(com.fasterxml.jackson.databind.SerializationFeature.WRITE_DATES_AS_TIMESTAMPS , false);
+            String jsonString = "";
+            try {
+                jsonString = mapper.writeValueAsString(initialPayload);
+            }catch(JsonProcessingException ex){
+                log.error("unable to Process the intial payload.");
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/service/MongoDBService.java]: Done

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/service/MongoDBService.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+                return false;
+            }
+            MongoCollection<Document> collection = MongoDBServiceHelper.getMongoCollection(this.mongoClient, this.shepherdConfiguration.getDataSourceDetails().getDb_name(), this.shepherdConfiguration.getDataSourceDetails().getCollection_name());
+
+            if (collection != null) {
+                final Document dbObjectInput = Document.parse(jsonString);
+                dbObjectInput.append("exec_id",executionId);
+                dbObjectInput.append("createdAt", new Date());
+                collection.insertOne(dbObjectInput);
+                log.debug("Object : " + jsonString);
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/service/MongoDBService.java]: done.

File: shepherd-client/configurations/dev-shepherd-configuration.yaml [hunk 10]
```diff
@@ -4,4 +4,17 @@ server :
   port : 8080
 
 headers :
-  Content-Type : application/json
\ No newline at end of file
+  Content-Type : application/json
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/configurations/dev-shepherd-configuration.yaml]: Yes, It should be inside shepherd-core.

[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/configurations/dev-shepherd-configuration.yaml]: changed.

File: shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/MongoDBHandler.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import com.mongodb.*;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class MongoDBHandler {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDBHandler.class);
+
+    public static String createMongoConnectionUri(DataSourceDetails dataSourceDetails) {
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/MongoDBHandler.java]: removed.

File: shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java [hunk 12]
```diff
@@ -36,13 +50,17 @@
     private ShepherdServerConfiguration shepherdServerDetails;
     private ShepherdClientHelper shepherdClientHelper;
 
+    private MongoClient mongoClient;
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java]: not needed

File: shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java [hunk 13]
```diff
@@ -198,4 +212,49 @@ public ShepherdResponse executeEndpoint(String clientName, String endpointName,
         }
 
     }
+
+    public boolean saveExecutionDetails(String executionID, String updatedInput) {
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java]: No need, removed

File: shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java [hunk 14]
```diff
@@ -198,4 +212,49 @@ public ShepherdResponse executeEndpoint(String clientName, String endpointName,
...
+                dbObjectUpdateInput.append("exec_id",executionID);
+                dbObjectUpdateInput.append("lastUpdatedDate",new Date());
+                UpdateResult updateResult = collection.replaceOne(dbObjectInput, dbObjectUpdateInput);
+                log.debug("updateDocument() :: database: " + this.shepherdServerDetails.getDataSourceDetails().getDb_name() + " and collection: " + this.shepherdServerDetails.getDataSourceDetails().getCollection_name()
+                        + " is document Updated :" + updateResult.wasAcknowledged());
+                boolean ack = updateResult.wasAcknowledged();
+                return ack;
+            }
+        } catch (MongoWriteException e) {
+            log.error(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java]: No need removed.

File: shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java [hunk 15]
```diff
@@ -198,4 +212,49 @@ public ShepherdResponse executeEndpoint(String clientName, String endpointName,
...
+        try
+        {
+            MongoCollection<Document> collection = MongoDBHandler.getMongoCollection(this.mongoClient, this.shepherdServerDetails.getDataSourceDetails().getDb_name(), this.shepherdServerDetails.getDataSourceDetails().getCollection_name());
+            if (collection != null) {
+                Document result = collection.find(new Document().append("exec_id",executionID)).first();
+                return result;
+            }
+
+        }catch(Exception ex){
+            ex.printStackTrace();
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/src/main/java/com/devsda/platform/shepherdclient/impl/ShepherdClient.java]: No need removed

File: shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/DocumentInputClass.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
+package com.devsda.platform.shepherdclient.impl;
+
+import java.util.Date;
+
+public class DocumentInputClass {
+
+    private String name;
+    private String payLoad;
```
[the remark this answers is not in the document]
Comment (JoooJooo) [reply shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/DocumentInputClass.java]: removved

File: shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/PayLoad.java (hunk 1 above)
Comment (JoooJooo) [reply shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/PayLoad.java]: removed, this was for testing purpose

File: shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/ShepherdClientTest.java (hunk 2 above)
Comment (JoooJooo) [reply shepherd-client/src/test/java/com/devsda/platform/shepherdclient/impl/ShepherdClientTest.java]: Alas this is also removed.

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java [hunk 17]
```diff
@@ -2,9 +2,13 @@
 
 import com.devsda.platform.shephardcore.constants.ShephardConstants;
 import com.devsda.platform.shephardcore.service.ExecuteWorkflowService;
+import com.devsda.platform.shephardcore.service.ExecutionDocumentService;
 import com.devsda.platform.shepherd.constants.ResourceName;
 import com.devsda.platform.shepherd.model.ExecuteWorkflowRequest;
 import com.google.inject.Inject;
+import com.mongodb.MongoWriteException;
```
Comment (devvsda) [inline shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java]: Unused imports.

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecutionDocumentResources.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Something went wrong " + e.getMessage()).build();
+        }
+    }
+
+    @Path("/{executionID}")
+    @PUT
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response updateExecutionDetails(@PathParam("executionID") Integer executionID,  String updatedExecutionDetails) throws Exception {
+
```
Comment (devvsda) [inline shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecutionDocumentResources.java]: We need to add logging.
We can prioritize later once we complete V1 cut pipeline.

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/service/ExecutionDocumentService.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+            String jsonString = "";
+            try {
+                jsonString = mapper.writeValueAsString(initialPayload);
+            }catch(JsonProcessingException e){
+                log.error(String.format("Unable to Process the initial payload for execution id : %s.", executionId), e);
+                return false;
+            }
+            MongoCollection<Document> collection = ExecutionDocumentServiceHelper.getMongoCollection(this.mongoClient, this.shepherdConfiguration.getDataSourceDetails().getDbname(), this.shepherdConfiguration.getDataSourceDetails().getCollectionname());
+
+            if (collection != null) {
```
Comment (devvsda) [inline shephard-core/src/main/java/com/devsda/platform/shephardcore/service/ExecutionDocumentService.java]: I think, we should block one field in doocument for Shepherd Metadata.

Name can be : "shepherd_metadata" : { "execution_id" : 12, "client_id" : 14, etc, etc}.

File: shepherd-client/pom.xml [hunk 20]
```diff
@@ -10,6 +10,13 @@
 
     <dependencies>
 
+       <!-- https://mvnrepository.com/artifact/org.mongodb/mongo-java-driver -->
```
Comment (devvsda) [inline shepherd-client/pom.xml]: Why this dependency here ?

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java (hunk 17 above)
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecuteWorkflowResources.java]: Removed

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecutionDocumentResources.java (hunk 18 above)
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/resources/ExecutionDocumentResources.java]: Sure.

File: shepherd-client/pom.xml (hunk 20 above)
Comment (JoooJooo) [reply shepherd-client/pom.xml]: removed

File: shephard-core/src/main/java/com/devsda/platform/shephardcore/service/ExecutionDocumentService.java (hunk 19 above)
Comment (JoooJooo) [reply shephard-core/src/main/java/com/devsda/platform/shephardcore/service/ExecutionDocumentService.java]: added TODO

**Pull request merged**